### PR TITLE
KIP-1071: Fix StickyTaskAssignor

### PR DESCRIPTION
**Pull request by aliehsaeedii**
Fixing Streams `StickyTaskAssignor` bug.

### CODE REVIEW

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/taskassignor/StickyTaskAssignor.java [hunk 1]
```diff
@@ -309,7 +299,14 @@ private void assignStandby(final Set<TaskId> standbyTasks, final int numStandbyR
 
                 // others
                 if (standby == null) {
-                    standby = findMemberWithLeastLoad(task, availableProcesses);
+                    final Set<Member> availableMembers = availableProcesses.stream()
+                        .flatMap(pId -> processIdToState.get(pId).memberToTaskCounts().keySet().stream()
+                            .map(mId -> new Member(pId, mId))).collect(Collectors.toSet());
```
Comment (cadonna) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/taskassignor/StickyTaskAssignor.java:304]: nit:
```suggestion
                    final Set<Member> availableMembers = availableProcesses.stream()
                        .flatMap(pId -> processIdToState.get(pId).memberToTaskCounts().keySet().stream()
                                .map(mId -> new Member(pId, mId)))
                        .collect(Collectors.toSet());
```

**Review by cadonna — APPROVED** [review]
LGTM!

Just one nit that we can change in a follow-up PR.

Thanks!

**Comment by aliehsaeedii** [comment]
> in a follow-up PR.

Thanks, @cadonna. What do you mean by `follow-up PR`?

**Comment by cadonna** [comment]
Sorry for not being clear enough!
We will still merge a couple of PRs on this dev branch. We can make the change I suggested in one of those future PRs. Or, if you prefer, you can push the change to this PR.

Actually, I still have a question about this PR.
Is it possible to add a unit test that verifies the fix you made?

**Review by bbejeck — APPROVED** [review]
@aliehsaeedii thanks for the PR this LGTM - for my own education could you point out in this PR which part is the fix?

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/taskassignor/StickyTaskAssignor.java [hunk 2]
```diff
@@ -226,6 +223,9 @@ private void maybeUpdateTasksPerMember(final int activeTasksNo) {
     }
 
     private Member findMemberWithLeastLoad(final Set<Member> members, TaskId taskId, final boolean returnSameMember) {
+        if (members == null || members.isEmpty()) {
+            return null;
+        }
```
Comment (aliehsaeedii) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/taskassignor/StickyTaskAssignor.java:228]: This is the fix.

**Comment by aliehsaeedii** [comment]
> @aliehsaeedii thanks for the PR this LGTM - for my own education could you point out in this PR which part is the fix?

Thanks @bbejeck. I left 2 comments to show the fix part.

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/taskassignor/StickyTaskAssignor.java [hunk 3]
```diff
@@ -309,7 +299,14 @@ private void assignStandby(final Set<TaskId> standbyTasks, final int numStandbyR
 
                 // others
                 if (standby == null) {
-                    standby = findMemberWithLeastLoad(task, availableProcesses);
+                    final Set<Member> availableMembers = availableProcesses.stream()
+                        .flatMap(pId -> processIdToState.get(pId).memberToTaskCounts().keySet().stream()
+                            .map(mId -> new Member(pId, mId))).collect(Collectors.toSet());
+                    standby = findMemberWithLeastLoad(availableMembers, task, false);
```
Comment (aliehsaeedii) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/taskassignor/StickyTaskAssignor.java:305]: fix part2

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/taskassignor/StickyTaskAssignor.java [hunk 4]
```diff
@@ -309,7 +299,14 @@ private void assignStandby(final Set<TaskId> standbyTasks, final int numStandbyR
...
+                    if (standby == null) {
+                        log.warn("{} Error in standby task assignment!", errorMessage);
+                        break;
+                    }
```
Comment (aliehsaeedii) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/taskassignor/StickyTaskAssignor.java:309]: fix part 2.

**Comment by aliehsaeedii** [comment]
> Sorry for not being clear enough! We will still merge a couple of PRs on this dev branch. We can make the change I suggested in one of those future PRs. Or, if you prefer, you can push the change to this PR.
> 
> Actually, I still have a question about this PR. Is it possible to add a unit test that verifies the fix you made?

Good point. Let's merge this PR and then I add a utest in the next PR.

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/taskassignor/StickyTaskAssignor.java (hunk 2 above)
Comment (bbejeck) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/taskassignor/StickyTaskAssignor.java:228]: Thanks @aliehsaeedii!

**Pull request merged**